File: solutions/typescript/clock/1/clock.ts
export class Clock {
  private static MINUTES_IN_HOUR = 60;
  private static HOURS_IN_DAY = 24;
  protected readonly _hour: number;
  protected readonly _minute: number;

  constructor(hour: number, minute = 0) {
    this._hour = this.getStandardizedHour(hour, minute);
    this._minute = this.getStandardizedMinute(minute);
  }

  public toString(): string {
    return this._hour.toString().padStart(2, "0") + ":" + this._minute.toString().padStart(2, "0");
  }

  public plus(minutes: number): Clock {
    return new Clock(this._hour, this._minute + minutes);
  }

  public minus(minutes: number): Clock {
    return new Clock(this._hour, this._minute - minutes);
  }

  public equals(other: Clock): boolean {
    return this._hour === other._hour && this._minute === other._minute;
  }

  private getStandardizedHour(hours: number, minutes: number): number {
    const additionalHour = Math.floor(minutes / Clock.MINUTES_IN_HOUR);
    let newHour = (hours + additionalHour) % Clock.HOURS_IN_DAY;

    if (newHour < 0)
      newHour = Clock.HOURS_IN_DAY + newHour;

    return newHour;
  }

  private getStandardizedMinute(minutes: number): number {
    let newMinutes = minutes % Clock.MINUTES_IN_HOUR;
    if (newMinutes < 0)
      newMinutes = Clock.MINUTES_IN_HOUR + newMinutes;

    return newMinutes;
  }
}

File: solutions/csharp/clock/1/Clock.cs
using System;

public class Clock : IEquatable<Clock>
{
    private readonly int _hours;
    private readonly int _minutes;

    public Clock(int hours, int minutes)
    {
        var totalMinutes = GetTotalMinutes(hours, minutes);
        _hours = GetNormalizedHours(totalMinutes);
        _minutes = GetNormalizedMinutes(totalMinutes);
        if (totalMinutes < 0 && _minutes == 0)
        {
            // Special case where we do not start back from 23:00
            _hours++;
        }
    }

    public Clock Add(int minutesToAdd) => new(_hours, _minutes + minutesToAdd);

    public Clock Subtract(int minutesToSubtract) => new(_hours, _minutes - minutesToSubtract);

    public bool Equals(Clock other)
    {
        if (other is null)
        {
            return false;
        }

        if (ReferenceEquals(this, other))
        {
            return true;
        }

        return _hours == other._hours && _minutes == other._minutes;
    }

    public override string ToString() => $"{GetPadding(_hours)}:{GetPadding(_minutes)}";

    private int GetTotalMinutes(int hours, int minutes) => hours * 60 + minutes;

    private int GetNormalizedHours(int totalMinutes)
    {
        var hours = totalMinutes / 60;
        if (totalMinutes < 0)
        {
            return GetWrapHours(hours);
        }
        return hours % 24;
    }

    private int GetWrapHours(int hours) => hours % 24 + 23;

    private int GetNormalizedMinutes(int totalMinutes)
    {
        var minutes = totalMinutes % 60;
        if (minutes < 0)
        {
            minutes = 60 + minutes;
        }
        return minutes;
    }

    private string GetPadding(int number) => number < 10 ? $"0{number}" : number.ToString();
}
